### PR TITLE
resolves #45 use background_color method to access background_color

### LINKED ISSUE #45: Background color of alternating row may bleed through on rowspan
If one of the two row colors is nil (not set), the color of the alternate row will bleed through rows that are part of a rowspan.

The following image demonstrates the error:

rowspan-bleed

Below the Hardware column header is a single cell that spans the remainder of the rows. Therefore, there should be no shaded body rows in the left column.

The problem is that the logic that assigns cell background colors doesn't skip SpanDummy cells. Therefore, the background of these cells get "painted" before the rowspan cell gets painted. If the background color of the rowspan cell is nil, it can't repaint over these cells.

### PR DESCRIPTION
**Pull request by mojavelinux**
- use background_color method to access background_color to honor rowspan
- add test for rowspan cell background color
- alternating row colors will no longer bleed through rowspan

### CODE REVIEW

**Comment by packetmonkey** [comment]
I think I'm good with the change, although I'm not very familiar with the codebase.

I would like to see the change to `draw_background` use a guard clause similar to

```ruby
      def draw_background(pt)
        return unless background_color

          @pdf.mask(:fill_color) do
            @pdf.fill_color background_color
            @pdf.fill_rectangle pt, width, height
          end
      end
``` 

Also it looks like the travis configs need to be updated, I'll try to handle that really quick.

Can you also please add an entry into the changeling referencing the fix?

Thanks.

**Comment by packetmonkey** [comment]
I attempted to update the supported ruby versions in PR #47 but jruby was failing, and i'm not sure how to debug that. I'm hesitant to not support jruby because we do support it in prawn but I also don't necessarily want to merge a PR with a failing test suite.

Thoughts?

**Comment by mojavelinux** [comment]
I'll dive in. Stay tuned.

**Comment by mojavelinux** [comment]
I can provide some additional context that should help with reviewing the change. The `background_color `reader is [overridden](https://github.com/prawnpdf/prawn-table/blob/master/lib/prawn/table/cell/span_dummy.rb#L73-L75) in the SpanDummy class to delegate the call to the master cell. Currently, the code is bypassing this override by directly accessing `@background_color`. The proposed change allows the method override to be used (and hence using the background color of the master cell).

I don't really understand how the code you suggested differs from what I have, other than a style change. I'm pretty sure they are effectively the same. But I'm happy to change it.

> I attempted to update the supported ruby versions in PR #47 but jruby was failing

JRuby 1.7.19 (the version used when you request the jruby keyword in RVM) is not a valid Ruby 2.0 implementation. While it has a Ruby 2 mode, that mode is known to be broken in several ways...one of which is affecting the test suite. If you want to test with JRuby, you have to use JRuby 9.0.0.0. You can enable JRuby 9.0.0.0 in RVM by using either `jruby-9.0.0.0.pre1` or simply `jruby9k`. I tested with JRuby 9.0.0.0 and all the tests pass.

By not supporting Ruby 1.9.3, you are effectively not supporting JRuby 1.7.

> I'm hesitant to not support jruby because we do support it in prawn

JRuby support is essential for us in Asciidoctor or else Asciidoctor PDF (which is based on Prawn) wouldn't work in AsciidoctorJ (Asciidoctor on the JVM).

**Comment by mojavelinux** [comment]
Let me know when you are ready for me to rebase.

**Comment by practicingruby** [comment]
I'd assume that what @packetmonkey meant was that he wouldn't want to break JRuby support because it is supported in Prawn. Switching to 2.0.0+ did essentially limit the support we offer for JRuby 1.7, but Prawn's test suite does pass on JRuby 1.7's 2.0 mode.

This is a situation where our hands are forced, because not even security patches are released for Ruby 1.9.x anymore, it's been officially dead-ended by the Ruby core team. Hopefully JRuby 9000 will stabilize in the near future, and we can just recommend others to use it. JRuby support is a high priority (in my now-retired opinion), but we can only support Ruby as far back as the core team does, since that's still our main userbase.

**Comment by mojavelinux** [comment]
I totally understand what you are saying, @practicingruby. Thanks for providing that insight.

I've been using JRuby 9K fairly regularly and I can say that it's at least as good as JRuby 1.7 at this point. I've also heard Charles and Tom confirm that JRuby 9K is passing the same rigor of tests that JRuby 1.7 goes through. So I'd say it's either on the verge of being stable or it already is...and will continue to get even better.

I agree that the situation with JRuby is tricky because Ruby 1.9 is clearly dead, but JRuby 1.7 never made it to Ruby 2. However, I feel pretty confident that JRuby 9K is going to be a turning point for JRuby in keeping up with MRI. Given that Red Hat now has its weight behind JRuby certainly helps add certainty to its future.

**Comment by practicingruby** [comment]
According to the [JRuby Roadmap](https://github.com/jruby/jruby/wiki/Roadmap), JRuby 9.0.0.0 RC1 was released on June 10, and a final release should drop soon. If that's the case, I'd say we just formally switch to supporting JRuby 9K when that happens, if not sooner.

**Comment by mojavelinux** [comment]
:+1:

File: lib/prawn/table/cell.rb [hunk 1]
```diff
@@ -710,9 +710,9 @@ def border_left_line=(val)
       # Draws the cell's background color.
       #
       def draw_background(pt)
-        if defined?(@background_color) && @background_color
+        if (bg_color = background_color)
```
Comment (johnnyshields) [inline lib/prawn/table/cell.rb]: What is the purpose of the extra variable assignment here?

Comment (mojavelinux) [inline lib/prawn/table/cell.rb]: To avoid calling the method a second time inside the conditional block. However, I realize this optimization is probably unnecessarily (and, in some cases, a deoptimization). I'll update it once I get the ping from @packetmonkey.

**Comment by packetmonkey** [comment]
> I don't really understand how the code you suggested differs from what I have, other than a style change. I'm pretty sure they are effectively the same. But I'm happy to change it.

You are correct, it's entirely a style change. I prefer this style because it reduces nesting and separates the validation of should this method execute from the actual execution, making further refactoring easier.

I also agree that JRuby is important and shouldn't be dropped, thanks @practicingruby for helping to clarify. That's what I get for trying to slip in a quick pull request review before bed :)

Given how this PR doesn't break the build, as it was already broken, I'm leaning towards merging the pull request after the style change is made and a changelog entry is created, but not cutting a release of prawn-table until the build is passing on all supported ruby versions.

@mojavelinux Given you have experience with JRuby would it be possible for you to take a swing at getting the test suite passing under 2.0 on JRuby?

**Comment by practicingruby** [comment]
Another thing to consider here because `prawn-table` is not technically subject to the same maintenance cycle as Prawn itself...

* Start building against JRuby 9000 immediately here, and require JRuby 9000 for prawn-table (don't even bother w. JRuby 1.7)

* Add JRuby 9000 to the main Prawn repository's CI as an allowable failure, but aim to fix any incompatibilities

* Once any issues are fixed and a final release of JRuby 9.0.0.0 happens, switch to it in Prawn proper and drop JRuby 1.7 support.

In other words... we don't need to release a prawn-table release if we require people to use JRuby 9.0.0.0, but in order to do that we should at least make sure Prawn itself is passing in CI.

**Comment by packetmonkey** [comment]
I like that suggestion a lot. It allows us to stay current with JRuby and will bring eventual consistency with the supported ruby versions across all the prawn projects.

If nobody has a problem with this I think that is what we should do.

/cc @hbrandl just in case he has been missing this discussion.

**Comment by practicingruby** [comment]
Ultimately it's @hbrandl's call here, because he's the semi-official maintainer of `prawn-table`.

**Comment by mojavelinux** [comment]
> You are correct, it's entirely a style change. I prefer this style because it reduces nesting and separates the validation of should this method execute from the actual execution, making further refactoring easier.

Thanks for the explanation. That makes complete sense. I'll update the code asap.

**Comment by mojavelinux** [comment]
:+1: to embracing JRuby 9.0.0.0. To clarify, the test suite does run on JRuby 9.0.0.0. The RVM on Travis CI only has JRuby 9.0.0.0.pre1 available, despite the fact that rc1 is released as @practicingruby pointed out. Probably best just to add the entry as `jruby9k` (the alias in RVM for JRuby 9.0.0.0).

**Comment by mojavelinux** [comment]
The test suite in Prawn core also passes on JRuby 9.0.0.0.

**Comment by mojavelinux** [comment]
I also recommend changing the JRUBY_OPTS environment variable as follows:

```
env:
  JRUBY_OPTS=--dev
```

The `--dev` flag tries to make JRuby run as fast as possible for small operations (like a test suite). In my tests, it cut the duration of the run in half. Once we switch to `jruby9k`, the Ruby mode does not have to be, nor should it be, set explicitly.

**Comment by mojavelinux** [comment]
Pull request updated!

**Comment by packetmonkey** [comment]
I just merged #47 so you should be able to rebase and we should see a passing test suite then we can hopefully get this merged in for you @mojavelinux

**Comment by mojavelinux** [comment]
Rebase complete! Awaiting results.

**Comment by packetmonkey** [comment]
And merged! Thanks

**Comment by mojavelinux** [comment]
Thank you!
